File: STK500.py
#!/usr/bin/python

#import libraries
import time
import bluetooth as bt

# define a rfcomm bluetooth connection
sock = bt.BluetoothSocket(bt.RFCOMM)
# connect to the bluetooth device
sock.connect(("98:D3:31:B1:CA:BF", 1))

# get in sync with the AVR
for i in range(5):
  print "syncing"
  sock.send('\x30') # STK_GET_SYNC
  sock.send('\x20') # STK_CRC_EOP
  time.sleep(0.05)

# receive sync ack
print "receiving sync ack"
insync = sock.recv(1) # STK_INSYNC
ok = sock.recv(1) # STK_OK

# check received ack
if insync == '\x14' and ok == '\x10':
  print "insync"

# get the MAJOR version of the bootloader
print "getting the MAJOR version of the bootloader"
sock.send('\x41') # STK_GET_PARAMETER
sock.send('\x81') # STK_SW_MAJOR
sock.send('\x20') # SYNC_CRC_EOP
time.sleep(0.05)

# receive bootlader MAJOR version
print "receiving bootloader MAJOR version"
insync = sock.recv(1) # STK_INSYNC
major = sock.recv(1) # STK_SW_MJAOR
ok = sock.recv(1) # STK_OK

# check received sync ack
if insync == '\x14' and ok == '\x10':
  print "insync"

# get the MINOR version of the bootloader
print "getting the MINOR version of the bootloader"
sock.send('\x41') # STK_GET_PARAMETER
sock.send('\x82') # STK_SW_MINOR
sock.send('\x20') # SYNC_CRC_EOP
time.sleep(0.05)

# receive bootlader MINOR version
print "receiving bootloader MINOR version"
insync = sock.recv(1) # STK_INSYNC
minor = sock.recv(1) # STK_SW_MINOR
ok = sock.recv(1) # STK_OK

# check received sync ack
if insync == '\x14' and ok == '\x10':
  print "insync"

print "bootloader version %s.%s" % (ord(major), ord(minor))

# enter programming mode
print "entering programming mode"
sock.send('\x50') # STK_ENTER_PROGMODE
sock.send('\x20') # SYNC_CRC_EOP
time.sleep(0.05)

# receive sync ack
print "receiving sync ack"
insync = sock.recv(1) # STK_INSYNC
ok = sock.recv(1) # STK_OK

# check received sync ack
if insync == '\x14' and ok == '\x10':
  print "insync"

# get device signature
print "getting device signature"
sock.send('\x75') # STK_READ_SIGN
sock.send('\x20') # SYNC_CRC_EOP

# receive device signature
print "receiving device signature"
insync = sock.recv(1) # STK_INSYNC
signature = sock.recv(3) # device
ok = sock.recv(1) # STK_OK

# check received sync ack
if insync == '\x14' and ok == '\x10':
  print "insync"

print "device signature %s-%s-%s" % (ord(signature[0]), ord(signature[1]), ord(signature[2]))

# start with page address 0
address = 0

# open the hex file
program = open("main.hex", "rb")

while True:
  # calculate page address
  laddress = chr(address % 256)
  haddress = chr(address / 256)
  address += 64

  # load page address
  print "loading page address"
  sock.send('\x55') # STK_LOAD_ADDRESS
  sock.send(laddress)
  sock.send(haddress)
  sock.send('\x20') # SYNC_CRC_EOP
  #time.sleep(0.01)

  # receive sync ack
  print "receiving sync ack"
  insync = sock.recv(1) # STK_INSYNC
  ok = sock.recv(1) # STK_OK

  # check received sync ack
  if insync == '\x14' and ok == '\x10':
    print "insync"

  data = ""
  # 16 bytes in a line 16 * 8 = 128
  for i in range(8):
    # just take the data part
    data += program.readline()[9:][:-4]

  size = chr(len(data)/2)
  print "sending program page to write", ord(size)
  sock.send('\x64') # STK_PROGRAM_PAGE
  sock.send('\x00') # page size
  sock.send(size) # page size
  sock.send('\x46') # flash memory, 'F'
  # while data left
  while data:
    # assemble a byte and send it
    sock.send(chr(int(data[:2], 16)))
    # chop of sent data
    data = data[2:]
  sock.send('\x20') # SYNC_CRC_EOP
  #time.sleep(0.01)

  # receive sync ack
  print "receiving sync ack"
  insync = sock.recv(1) # STK_INSYNC
  ok = sock.recv(1) # STK_OK

  # check received sync ack
  if insync == '\x14' and ok == '\x10':
    print "insync"

  # when the whole program was uploaded
  if size != '\x80':
    break

# close the hex file
program.close()

# load page address
#print "loading page address"
#sock.send('\x55') # STK_LOAD_ADDRESS
#sock.send('\x00')
#sock.send('\x00')
#sock.send('\x20') # SYNC_CRC_EOP
#time.sleep(0.05)

# receive sync ack
#print "receiving sync ack"
#insync = sock.recv(1) # STK_INSYNC
#ok = sock.recv(1) # STK_OK

# check received sync ack
#if insync == '\x14' and ok == '\x10':
#  print "insync"

# send read program page
#print "sending program page to read"
#sock.send('\x74') # STK_READ_PAGE
#sock.send('\x00') # page size
#sock.send('\x80') # page size
#sock.send('\x46') # flash memory, 'F'
#sock.send('\x20') # SYNC_CRC_EOP
#time.sleep(0.05)

#print len(sock.recv(128))

# leave programming mode
print "leaving programming mode"
sock.send('\x51') # STK_LEAVE_PROGMODE
sock.send('\x20') # SYNC_CRC_EOP
time.sleep(0.05)

# receive sync ack
print "receiving sync ack"
insync = sock.recv(1) # STK_INSYNC
ok = sock.recv(1) # STK_OK

# check received sync ack
if insync == '\x14' and ok == '\x10':
  print "insync"

# close the bluetooth connection
sock.close()

File: STK500.java
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.io.FileNotFoundException;

public class STK500 {
    private OutputStream outStream = null;
    private InputStream inputStream = null;

    private static void openStreams() {
        try {
            /* TODO: implement opening bluetooth */
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void closeStreams() {
        try {
            /* TODO: implement closing bluetooth */
            outStream.close();
            inputStream.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static byte[] readProgram() {
        FileInputStream fis = null;
        File file = new File("/sdcard/sumorobot/main.hex");

        // every line, except last one, has has 45 bytes (including \r\n)
        int programLines = (int) Math.ceil(file.length() / 45.0);
        // every line has 32 bytes of program data (excluding checksums, addresses, etc.)
        int unusedBytes = 45 - 32;
        // calculate program length according to program lines and unused bytes
        int programLength = (int) file.length() - (programLines * unusedBytes);
        // the actualy program data is half the size, as the hex file represents hex data in individual chars
        programLength /= 2;
        // create a byte array with the program length
        byte[] program = new byte[programLength];

        try {
            // open the file stream
            log.logcat("opening hex file", "d");
            fis = new FileInputStream(file);
            log.logcat("Total program size (in bytes) : " + programLength, "d");
            log.logcat("Total file size to read (in bytes) : " + fis.available(), "d");

            int content;
            int lineIndex = 0;
            int lineNumber = 1;
            int programIndex = 0;
            char[] line = new char[45];
            // read the file byte by byte
            while ((content = fis.read()) != -1) {
                // append byte to the line
                line[lineIndex++] = (char) content;
                // when the line is complete
                if (content == 10) {
                    // take only the actual program data form the line
                    for (int index = 9; index < lineIndex - 4; index += 2) {
                        // convert hexadecimals represented as chars into bytes
                        program[programIndex++] = Integer.decode("0x" + line[index] + line[index+1]).byteValue();
                    }
                    // start a new line
                    lineIndex = 0;
                }
            }
        } catch (IOException e) {
            log.logcat("reading hex failed: " + e.getMessage(), "d");
        } finally {
            try {
                if (fis != null)
                    fis.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        return program;
    }

    public static void main (String[] args) {
        byte[] program = readProgram();
        System.out.println("program length: " + program.length);

        System.out.println("opening streams");
        openStreams();

        System.out.println("syncing");
        for (int i = 0; i < 5; i++) {
            outStream.write(0x30);
            outStream.write(0x20);
            try {Thread.sleep(50);} catch (InterruptedException e) {e.printStackTrace();}
        }

        System.out.println("waiting for response");
        int insync = inputStream.read();
        int ok = inputStream.read();
        if (insync == 0x14 && ok == 0x10) {
            System.out.println("insync");
        }

        System.out.println("reading major version");
        outStream.write(0x41);
        outStream.write(0x81);
        outStream.write(0x20);
        try {Thread.sleep(50);} catch (InterruptedException e) {e.printStackTrace();}

        System.out.println("waiting for response");
        insync = inputStream.read();
        int major = inputStream.read();
        ok = inputStream.read();
        if (insync == 0x14 && ok == 0x10) {
            System.out.println("insync", "d");
        }

        System.out.println("reading minor version");
        outStream.write(0x41);
        outStream.write(0x82);
        outStream.write(0x20);
        try {Thread.sleep(50);} catch (InterruptedException e) {e.printStackTrace();}

        System.out.println("waiting for response");
        insync = inputStream.read();
        int minor = inputStream.read();
        ok = inputStream.read();
        if (insync == 0x14 && ok == 0x10) {
            System.out.println("insync");
        }

        System.out.println("version: " + major + "." + minor);

        System.out.println("entering programming mode");
        outStream.write(0x50);
        outStream.write(0x20);
        try {Thread.sleep(50);} catch (InterruptedException e) {e.printStackTrace();}

        System.out.println("waiting for response");
        insync = inputStream.read();
        ok = inputStream.read();
        if (insync == 0x14 && ok == 0x10) {
            System.out.println("insync");
        }

        System.out.println("getting device signature");
        outStream.write(0x75);
        outStream.write(0x20);
        try {Thread.sleep(50);} catch (InterruptedException e) {e.printStackTrace();}

        System.out.println("waiting for response");
        insync = inputStream.read();
        byte [] signature = new byte[3];
        inputStream.read(signature, 0, 3);
        ok = inputStream.read();
        if (insync == 0x14 && ok == 0x10) {
            System.out.println("insync");
        }

        System.out.println("signature: " + signature[0] + "." + signature[1] + "." + signature[2]);

        int size = 0;
        int address = 0;
        int programIndex = 0;
        while (true) {
            int laddress = address % 256;
            int haddress = address / 256;
            address += 64;

            System.out.println("loading page address");
            outStream.write(0x55);
            outStream.write(laddress);
            outStream.write(haddress);
            outStream.write(0x20);
            try {Thread.sleep(50);} catch (InterruptedException e) {e.printStackTrace();}

            System.out.println("waiting for response");
            insync = inputStream.read();
            ok = inputStream.read();
            if (insync == 0x14 && ok == 0x10) {
                System.out.println("insync");
            }

            if (program.length - programIndex < 128) {
                size = program.length - programIndex;
            } else {
                size = 128;
            }
            System.out.println("programming page size: " + size + " haddress: " + haddress + " laddress: " + laddress);
            outStream.write(0x64);
            outStream.write(0x00);
            outStream.write(size);
            outStream.write(0x46);
            for (int i = 0; i < size; i++) {
                outStream.write(program[programIndex++]);
            }
            outStream.write(0x20);
            try {Thread.sleep(50);} catch (InterruptedException e) {e.printStackTrace();}

            System.out.println("receiving sync ack");
            insync = inputStream.read();
            ok = inputStream.read();

            if (insync == 0x14 && ok == 0x10) {
                System.out.println("insync");
            }

            if (size != 0x80) {
                break;
            }
        }
        System.out.println("program index: " + programIndex);

        System.out.println("leaving programming mode");
        outStream.write(0x51);
        outStream.write(0x20);
        try {Thread.sleep(50);} catch (InterruptedException e) {e.printStackTrace();}

        System.out.println("receiving sync ack");
        insync = inputStream.read();
        ok = inputStream.read();

        if (insync == 0x14 && ok == 0x10) {
            System.out.println("insync");
        }

        System.out.println("closing streams");
        closeStreams();
    }
}